backend/embedding/main.py

```diff
@@ -169,44 +190,77 @@ def main(
-                sources,
-                embeddings,
-                chunk_size,
-                chunk_overlap,
-            )
-        if len(sitemap_urls) > 0:
-            for sitemap_url in sitemap_urls:
-                raise NotImplementedError()
-        if len(filenames) > 0:
-            for filename in filenames:
+        with multiprocessing.Manager() as manager:
+            contents: ListProxy = manager.list()
+            sources: ListProxy = manager.list()
+            embeddings: ListProxy = manager.list()
+
+            if len(source_urls) > 0:
                 embed(
-                    S3FileLoader(
-                        bucket=DOCUMENT_BUCKET,
-                        key=compose_upload_document_s3_path(user_id, bot_id, filename),
-                    ),
+                    UrlLoader(source_urls),
                     contents,
                     sources,
                     embeddings,
                     chunk_size,
                     chunk_overlap,
                 )
+            if len(sitemap_urls) > 0:
+                for sitemap_url in sitemap_urls:
+                    raise NotImplementedError()
+            if len(filenames) > 0:
+                with multiprocessing.Pool(processes=None) as pool:
+                    futures = [
+                        pool.apply_async(
+                            embed,
+                            args=(
+                                S3FileLoader(
+                                    bucket=DOCUMENT_BUCKET,
+                                    key=compose_upload_document_s3_path(
+                                        user_id, bot_id, filename
+                                    ),
+                                    enable_partition_pdf=enable_partition_pdf,
+                                ),
+                                contents,
+                                sources,
+                                embeddings,
+                                chunk_size,
+                                chunk_overlap,
+                            ),
+                        )
+                        for filename in filenames
+                    ]
+                    for future in futures:
+                        future.get()
+
+                logger.info(f"Number of chunks: {len(contents)}")
+
+                # Insert records into postgres
+                insert_to_postgres(bot_id, contents, sources, embeddings)
+    except Exception as e:
+        logger.error("[ERROR] Failed to embed.")
+        logger.error(e)
+        update_sync_status(
+            user_id,
+            bot_id,
+            "FAILED",
+            f"{e}",
+            exec_id,
+        )
+        return
 
-        print(f"Number of chunks: {len(contents)}")
-
-        # Insert records into postgres
-        insert_to_postgres(bot_id, contents, sources, embeddings)
+    try:
         status_reason = "Successfully inserted to vector store."
+        update_sync_status(
```

Comment:
as well as the above discussion, I'd like to know the reason why added this err handling